test/karma/spec/loginFlow.js

```diff
@@ -0,0 +1,263 @@
+      expect(url.searchParams.get('redirect_uri')).toBe(REDIRECT_URI);
+      expect(url.searchParams.get('response_type')).toBe('code');
+      expect(url.searchParams.get('response_mode')).toBe('fragment');
+      expect(url.searchParams.get('scope')).toBe('openid email');
+      expect(url.searchParams.get('state')).toBeTruthy();
+      expect(url.searchParams.get('nonce')).toBeTruthy();
+      expect(url.searchParams.get('code_challenge')).toBe(CODE_CHALLENGE);
+      return url;
+    })
+    .then(function(url) {
+      jasmine.Ajax.requests.reset();
+
+      // Now we handle the redirect & hit /token
+      const state = url.searchParams.get('state');
+      const pathname = `${CALLBACK_PATH}#code=${AUTHORIZATION_CODE}&state=${state}`;
+      const tokenResponse = {
+        'access_token': ACCESS_TOKEN,
+        'id_token': ID_TOKEN,
+        'nonce': NONCE,
+        'expires_in': 1000
+      };
+      jasmine.Ajax.stubRequest(
+        /.*v1\/token/
+      ).andReturn({
+        status: 200,
+        responseText: JSON.stringify(tokenResponse)
+      });
+
+      return bootstrap({}, pathname)
+      .then(function() {
+
+        expect(getLocation).toHaveBeenCalled();
+
+        // removeHash
+        expect(getHistory).toHaveBeenCalled();
+        expect(getDocument).toHaveBeenCalled(); 
+        expect(_history.replaceState).toHaveBeenCalledWith(null, DOC_TITLE, CALLBACK_PATH);
+
+        // Validate POST request to /token
+        var request = jasmine.Ajax.requests.first();
+        expect(request.url).toBe(`${ISSUER}/oauth2/v1/token`);
+        expect(request.method).toBe('POST');
+        expect(request.requestHeaders['content-type']).toBe('application/x-www-form-urlencoded');
+        expect(request.withCredentials).toBe(false);
+
+        // Decode request params
+        var params = {}
+        request.params.split('&').forEach(function(str) {
+          var pair = str.split('=');
+          params[pair[0]] = decodeURIComponent(pair[1]);
+        });
+        expect(params['client_id']).toBe(CLIENT_ID);
+        expect(params['redirect_uri']).toBe(REDIRECT_URI);
+        expect(params['grant_type']).toBe('authorization_code');
+        expect(params['code']).toBe(AUTHORIZATION_CODE);
+        expect(params['code_verifier']).toBe(CODE_VERIFIER);
+
+        // Validate claims were processed by app
+        var $claims = $app.find('#claims');
+        expect($claims.length).toBe(1);
+      });
+    });
+  });
+
+  it('throws for invalid code_challenge_method', function() {
+    // First hit /authorize
+    return bootstrap()
+    .then(function(app) {
+      mockWellKnown();
+      return app.loginRedirect({
+        grantType: 'authorization_code',
+        nonce: NONCE,
+        codeVerifier: CODE_VERIFIER,
+        codeChallengeMethod: 'invalid'
+      });
+    }).catch(function(e) {
+      expect(e instanceof AuthSdkError).toBe(true);
+      expect(e.message).toBe('Invalid code_challenge_method');
+    });
+  });
+
```

Comment:
Lets also add tests for:
1. Invalid `CODE_VERIFIER`
2. Fails when `crypto.subtle` isn't available (IE11)
3. Silent token renewal (open up an iframe leveraging the Okta session)